lib/jobs/ssh-job.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016, EMC, Inc.
+
+'use strict';
+
+var di = require('di');
+module.exports = sshJobFactory;
+
+di.annotate(sshJobFactory, new di.Provide('Job.Ssh'));
+di.annotate(sshJobFactory, new di.Inject(
+    'Job.Base',
+    'JobUtils.CommandParser',
+    'Util',
+    'Logger',
+    'Assert',
+    'Promise',
+    'Services.Waterline',
+    'Services.Encryption',
+    'ssh',
+    'JobUtils.Commands',
+    '_'
+));
+
+function sshJobFactory(
+    BaseJob,
+    parser,
+    util,
+    Logger,
+    assert,
+    Promise,
+    waterline,
+    cryptService,
+    ssh,
+    CommandUtil,
+    _
+) {
+    var logger = Logger.initialize(sshJobFactory);
+    var commandUtil;
+    function SshJob(options, context, taskId) {
+        SshJob.super_.call(this, logger, options, context, taskId);
+        assert.string(this.context.target);
+        this.nodeId = this.context.target;
+        commandUtil = new CommandUtil(this.nodeId);
+        this.commands = commandUtil.buildCommands(options.commands);
+        assert.arrayOfObject(this.commands);
+    }
+    util.inherits(SshJob, BaseJob);
+
+    SshJob.prototype._run = function run() {
+        var self = this;
+        return waterline.nodes.needByIdentifier(self.nodeId)
+        .then(function(node) {
+            return Promise.reduce(self.commands, function(results, commandData) {
+                return self.sshExec(commandData, node.sshSettings, new ssh.Client())
+                .then(function(result) {
+                    return results.concat([result]);
+                });
+            }, []);
+        })
+        .then(commandUtil.parseResponse.bind(commandUtil))
+        .spread(commandUtil.catalogParsedTasks.bind(commandUtil))
+        .spread(function() {
+            self._done();
+        })
+        .catch(self._done.bind(self));
+    };
+
+    SshJob.prototype.sshExec = function(cmdObj, sshSettings, sshClient) {
+        return new Promise(function(resolve, reject) {
+            if(cmdObj.timeout) {
+                setTimeout(function() {
+                    var seconds = cmdObj.timeout / 1000;
+                    reject(new Error('The remote operation timed out after '+
+                                 seconds + ' seconds'));
+                }, cmdObj.timeout);
+            }
+            var ssh = sshClient;
+            ssh.on('ready', function() {
+                ssh.exec(cmdObj.cmd, function(err, stream) {
+                    if (err) { reject(err); }
+                    stream.on('close', function(code) {
```

Comment:
Are we going to get fatal errors if we don't handle 'error' on this stream object?